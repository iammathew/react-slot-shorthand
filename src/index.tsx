import React from 'react';

export type PrimitiveShorthand =
  | number
  | string
  | boolean
  | React.ReactElement
  | never;

export type SlotsType = Record<
  string,
  {
    component: React.ComponentType<any>;
    shorthand: PrimitiveShorthand;
    optional: true | false;
  }
>;

export interface RenderPropsSlot<C extends React.ComponentType<any>> {
  children: (Component: C, props: React.ComponentProps<C>) => React.ReactNode;
}

function isRenderProps<C extends React.ComponentType<any>>(
  props: any
): props is RenderPropsSlot<C> {
  return props.children != null && typeof props.children === 'function';
}

export type PropsWithSlots<Props, Slots extends SlotsType> = Props &
  {
    [Property in keyof Slots]: Slots[Property]['optional'] extends true
      ?
          | RenderPropsSlot<Slots[Property]['component']>
          | (React.ComponentProps<Slots[Property]['component']> &
              Partial<RenderPropsSlot<Slots[Property]['component']>>)
          | Slots[Property]['shorthand']
          | null
          | undefined
      :
          | RenderPropsSlot<Slots[Property]['component']>
          | (React.ComponentProps<Slots[Property]['component']> &
              Partial<RenderPropsSlot<Slots[Property]['component']>>)
          | Slots[Property]['shorthand']
          | null;
  };

function Text() {
  return <p></p>;
}

export type Test = PropsWithSlots<
  {},
  {
    text?: {
      component: typeof Text;
      shorthand: React.ReactElement;
      optional: true;
    };
  }
>;

const test: Test = {};

export function useSlot<Component extends React.ComponentType<any>>(
  component: Component,
  props:
    | RenderPropsSlot<Component>
    | (React.ComponentProps<Component> & Partial<RenderPropsSlot<Component>>)
    | PrimitiveShorthand
    | null
    | undefined,
  options: {
    primitiveRemap:
      | keyof React.ComponentProps<Component>
      | 'display'
      | 'replace';
    defaultProps: React.ComponentProps<Component>;
  }
): () => React.ReactNode {
  if (props === null) {
    return () => null;
  }
  let appliedProps: React.ComponentProps<Component>;
  if (
    typeof props === 'string' ||
    typeof props === 'number' ||
    typeof props === 'boolean' ||
    React.isValidElement(props)
  ) {
    if (options.primitiveRemap === 'display') {
      if (props === false) {
        return () => {
          return null;
        };
      } else {
        appliedProps = options.defaultProps;
      }
    } else if (options.primitiveRemap === 'replace') {
      return () => {
        return props;
      };
    } else {
      appliedProps = {
        [options.primitiveRemap]: props,
      } as React.ComponentProps<Component>;
    }
  } else if (isRenderProps<Component>(props)) {
    const { children, ...restProps } = props;
    return () => {
      return props.children(
        component,
        restProps as React.ComponentProps<Component>
      );
    };
  } else if (props === undefined) {
    appliedProps = options.defaultProps;
  } else {
    appliedProps = props;
  }
  const Component = component;
  return () => {
    return <Component {...appliedProps} />;
  };
}
