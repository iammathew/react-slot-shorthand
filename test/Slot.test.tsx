import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { useSlot, PropsWithSlots } from '../src';

function Text({ body }: { body: string }) {
  return <p>{body}</p>;
}

function TextContainer({
  text,
}: PropsWithSlots<
  {},
  {
    text: {
      component: typeof Text;
      shorthand: string | React.ReactElement;
      optional: false;
    };
  }
>) {
  const renderText = useSlot(Text, text, {
    primitiveRemap: 'body',
    defaultProps: { body: 'haha' },
  });
  return <div>{renderText()}</div>;
}

function TextContainerOptional({
  text,
}: PropsWithSlots<
  {},
  {
    text: {
      component: typeof Text;
      shorthand: string | React.ReactElement | undefined;
      optional: true;
    };
  }
>) {
  const renderText = useSlot(Text, text, {
    primitiveRemap: 'body',
    defaultProps: { body: 'Default Text' },
  });
  return <div>{renderText()}</div>;
}

function TextContainerDisplay({
  text,
}: PropsWithSlots<
  {},
  { text: { component: typeof Text; shorthand: boolean; optional: false } }
>) {
  const renderText = useSlot(Text, text, {
    primitiveRemap: 'display',
    defaultProps: { body: 'Default Text' },
  });
  return <div>{renderText()}</div>;
}

function TextContainerReplace({
  text,
}: PropsWithSlots<
  {},
  {
    text: {
      component: typeof Text;
      shorthand: React.ReactElement;
      optional: false;
    };
  }
>) {
  const renderText = useSlot(Text, text, {
    primitiveRemap: 'replace',
    defaultProps: { body: 'Default props!' },
  });
  return <div>{renderText()}</div>;
}

describe('Slot TextContainer[body]', () => {
  it('renders correctly (string shorthand)', () => {
    const result = render(<TextContainer text="String Shorthand!" />);
    expect(result.getByText('String Shorthand!')).toBeInTheDocument();
    expect(result.container.firstChild).toMatchSnapshot();
  });

  it('renders correctly (react element shorthand)', () => {
    const result = render(<TextContainer text={<i>Element Shorthand!</i>} />);
    expect(result.getByText('Element Shorthand!')).toBeInTheDocument();
    expect(result.container.firstChild).toMatchSnapshot();
  });

  it('renders correctly (props)', () => {
    const result = render(<TextContainer text={{ body: 'Props used!' }} />);
    expect(result.getByText('Props used!')).toBeInTheDocument();
    expect(result.container.firstChild).toMatchSnapshot();
  });

  it('renders correctly (null)', () => {
    const result = render(<TextContainer text={null} />);
    expect(result.container.firstChild).toMatchSnapshot();
  });
});

describe('Slot TextContainerOptional[body]', () => {
  it('renders correctly (not defined)', () => {
    const result = render(<TextContainerOptional />);
    expect(result.getByText('Default Text!')).toBeInTheDocument();
    expect(result.container.firstChild).toMatchSnapshot();
  });
});
