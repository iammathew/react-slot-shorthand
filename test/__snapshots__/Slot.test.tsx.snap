// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot TextContainer[body] renders correctly (null) 1`] = `<div />`;

exports[`Slot TextContainer[body] renders correctly (props) 1`] = `
<div>
  <p>
    Props used!
  </p>
</div>
`;

exports[`Slot TextContainer[body] renders correctly (react element shorthand) 1`] = `
<div>
  <p>
    <i>
      Element Shorthand!
    </i>
  </p>
</div>
`;

exports[`Slot TextContainer[body] renders corretly (string shorthand) 1`] = `
<div>
  <p>
    String Shorthand!
  </p>
</div>
`;
